### PR TITLE
Add optional support for microsecond tracing

### PR DESCRIPTION
**Pull request by afbjorklund**
By setting $CCACHE_TRACEFILE, one can obtain a trace of an individual compile.

Setting a unique file per compile, there is a script to combine them all into one trace...

This trace can then be loaded into the `chrome://tracing` page of Chromium/Chrome:

ccache-tracing-miss

ccache-tracing-hit

Currently there are 7 main event categories: config, main, hash, manifest, cache, file, execute

### CODE REVIEW

File: src/ccache.c [hunk 1]
```diff
@@ -3323,23 +3411,31 @@ ccache(int argc, char *argv[])
...
 	}
 
+	MTR_BEGIN("main", "set_up_uncached_err");
 	set_up_uncached_err();
+	MTR_END("main", "set_up_uncached_err");
 
-	cc_log_argv("Command line: ", argv);
-	cc_log("Hostname: %s", get_hostname());
-	cc_log("Working directory: %s", get_current_working_dir());
+	if (!str_eq(conf->log_file, "")) {
```
Comment (jrosdahl) [inline src/ccache.c]: Why this? Is it because `get_hostname()` and `get_current_working_dir()` take time? If so, please move to a separate pull request.

File: src/util.c [hunk 2]
```diff
@@ -1688,3 +1688,14 @@ set_cloexec_flag(int fd)
 	(void)fd;
 #endif
 }
+
+double time_seconds(void)
+{
+#ifdef HAVE_GETTIMEOFDAY
+        struct timeval tv;
```
Comment (jrosdahl) [inline src/util.c]: Indent with tab (below as well).

File: src/minitrace.c [hunk 3]
```diff
@@ -0,0 +1,381 @@
+// minitrace
```
Comment (jrosdahl) [inline src/minitrace.c]: `src/minitrace.c` and `src/minitrace.h` should be added to `uncrustify_exclude_files` in `dev.mk.in`.

File: configure.ac [hunk 4]
```diff
@@ -126,6 +127,14 @@ else
     LIBS="$LIBS -lz"
 fi
 
+AC_ARG_ENABLE(tracing,
+  [AS_HELP_STRING([--enable-tracing],
+    [enable possibility to use tracing])])
```
[the remark this answers is not in the document]
Comment (afbjorklund) [reply configure.ac]: Done

File: doc/MANUAL.adoc [hunk 5]
```diff
@@ -1099,6 +1099,19 @@ problems and what may be done to increase the hit rate:
   <<_precompiled_headers,PRECOMPILED HEADERS>>.
 
 
+Tracing
```
[the remark this answers is not in the document]
Comment (afbjorklund) [reply doc/MANUAL.adoc]: It probably belongs in a developer manual, there could be other internal sections that could be lifted out as well.

File: doc/MANUAL.adoc [hunk 6]
```diff
@@ -1099,6 +1099,19 @@ problems and what may be done to increase the hit rate:
...
+
+In order to see what ccache is doing, it is possible to enable microsecond tracing.
+This needs to be done when compiling ccache, using the `--enable-tracing` feature.
+
+By setting *CCACHE_TRACEFILE*, one can obtain a JSON trace of an individual compile.
+This trace can then be loaded into the `chrome://tracing` page of Chromium/Chrome.
+
+The current event categories are: config, main, hash, manifest, cache, file, execute
+
+With a unique file per compile, there is a script to combine them all into one trace...
```
[the remark this answers is not in the document]
Comment (afbjorklund) [reply doc/MANUAL.adoc]: Done

File: src/ccache.c [hunk 7]
```diff
@@ -479,6 +479,13 @@ clean_up_internal_tempdir(void)
 static enum guessed_compiler
 guess_compiler(const char *path)
 {
+	if (str_eq(path, "/usr/bin/gcc")) {
```
[the remark this answers is not in the document]
Comment (afbjorklund) [reply src/ccache.c]: Yeah, they might have been a bit premature - and really only matter on the "cache hit (direct)" path anyway.

File: src/ccache.c (hunk 1 above)
Comment (afbjorklund) [reply src/ccache.c]: Yeah, same reason as above / the other one.

File: src/util.c (hunk 2 above)
Comment (afbjorklund) [reply src/util.c]: Done

**Comment by afbjorklund** [comment]
Travis seems to have given up on this

**Comment by afbjorklund** [comment]
It got annoying to have modify build systems, in order to set the $CCACHE_TRACEFILE.
So added an option to generate it dynamically, maybe it should even be the default ?

``` console
$ export CCACHE_TRACEFILE=""
$ ./ccache cc -c test.c
$ json_verify < test.o.ccache-trace 
JSON is valid
```

**Comment by afbjorklund** [comment]
Here's ccache tracing itself:
ccache-trace

**Comment by jrosdahl** [comment]
> It got annoying to have to modify build systems, in order to set $CCACHE_TRACEFILE.
> So added an option to generate it dynamically, maybe it should even be the default ?

Yes, I suggest making it the default and **only** mode. So if `CCACHE_INTERNAL_TRACE` is set, enable tracing?

File: doc/MANUAL.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (jrosdahl) [reply doc/MANUAL.adoc]: Yes, please move it to a developer manual.

> here could be other internal sections that could be lifted out as well.

Don't of any; the manual page has always been for the ccache user, not the ccache developer.

File: src/ccache.c [hunk 8]
```diff
@@ -3159,20 +3179,100 @@ create_initial_config_file(struct conf *conf, const char *path)
 	fclose(f);
 }
 
+#ifdef MTR_ENABLED
+static void *trace_id;
+#endif
+const char *trace_file;
+
+void trace_init(const char *json)
```
Comment (jrosdahl) [inline src/ccache.c]: How about putting all tracing stuff inside `#ifdef MTR_ENABLED`?

**Comment by jrosdahl** [comment]
@afbjorklund: Have you had time to have a look at my suggestions?

**Comment by jrosdahl** [comment]
@afbjorklund: I merged latest master into your branch for your convenience. However, I'm moving this feature out of 3.6 since it doesn't feel quite ready yet, mainly documentation. If you want to work on it near term, I'll happily move it back.

**Comment by afbjorklund** [comment]
Will merge `master` and try again

**Comment by afbjorklund** [comment]
Hmm, seems like support for old clang is broken on master ?

`error: macro name is a reserved identifier [-Werror,-Wreserved-id-macro]`

**Comment by jrosdahl** [comment]
> Interestingly, it had already been merged to my fork (guess that is what you wrote above).

Yes, GitHub by default allows maintainers to [push to the pull request branch](https://help.github.com/articles/allowing-changes-to-a-pull-request-branch-created-from-a-fork/). I intend to make use of that feature to make minor fixes directly on the pull request branch instead of making review change requests since they tend to stall very often.

**Comment by afbjorklund** [comment]
Confusing with all the extra warnings, only for clang ? Especially when happending to third parties.

**Comment by jrosdahl** [comment]
> Confusing with all the extra warnings, only for clang ? Especially when happending to third parties.

I agree. How about only passing `@more_warnings@` to `$(non_3pp_sources)`?

**Comment by jrosdahl** [comment]
> I agree. How about only passing @more_warnings@ to $(non_3pp_sources)?

Done in 39c531ff69a4.

File: doc/MANUAL.adoc (hunk 5 above)
[the remark this answers is not in the document]
Comment (afbjorklund) [reply doc/MANUAL.adoc]: Done

**Comment by jrosdahl** [comment]
Thanks!

**Comment by afbjorklund** [comment]
Unfortunately the tracefiles were lost in all the refactoring, they remain in the tmp directory...

Will submit a patch to fix it.
